synapseclient/client.py

```diff
@@ -1770,82 +1773,146 @@ def returnDict(destination):
                      'files': [None] if destination is None else [os.path.basename(destination)],
                      'cacheDir': None if destination is None else os.path.dirname(destination) }
 
-        if expected_md5 == '':
-            expected_md5 = None
+        return returnDict(self._download_with_retries(url, destination, file_handle_id, expected_md5))
 
-        ## download via http unless we're redirected to some other scheme
-        download_via_http = True
 
-        # We expect to be redirected to a signed S3 URL or externalURL
-        #The assumption is wrong - we always try to read either the outer or inner requests.get
-        #but sometimes we don't have something to read.  I.e. when the type is ftp at which point
-        #we still set the cache and filepath based on destination which is wrong because nothing was fetched
-        response = _with_retry(lambda: requests.get(url, headers=self._generateSignedHeaders(url), allow_redirects=False), verbose=self.debug, **STANDARD_RETRY_PARAMS)
+    def _download_with_retries(self, url, destination, file_handle_id=None, expected_md5=None, retries=5):
+        while retries > 0:
+            try:
```

Comment:
Reset retry counter if we have made progress since last loop.   Can be checked by keeping track of size of destination.